hypothesis-python/tests/django/toystore/test_given_models.py

```diff
@@ -135,6 +135,16 @@ def test_no_null_in_charfield(self, x):
         # regression test for #1045.  Company just has a convenient CharField.
         assert u'\x00' not in x.name
 
+    def test_foreign_key_primary(self):
+        # Regression test for #1307
+        strategy = models(
+            CompanyExtension,
+            company=models(Company),
+            self_modifying=just(2)
+        )
+        example = strategy.example()
```

Comment:
`.example()` isn't great for use in testing. Can you write this test to use `@given(...)`?

(Note that we lazily evaluate the strategies, so even if this stops working later the test runner will give us test failures instead of setup errors)

Reply:
I discussed my reasons for using `example()` on the issue (#1307). I think in order to hit the problem I'll need to draw twice using duplicate copies of a buffer, which I couldn't find a clean way of doing. I can definitely come up with something, but it seemed to involve more subversion of the API than seemed wise. I'll try and lash something together to illustrate what I mean.

Reply:
Hmm.  An idea that *might* work is to take the `binary()` strategy and then make a pair of `ConjectureData` objects out of it:

```python
@given(binary(min_size=20))
def test_foreign_key_primary(buffer):
    strategy = ...
    try:
        # Put a thing in the database, with determined values
        ConjectureData.for_buffer(buffer).draw(strategy)
    except BaseException:
        reject()  # whoops, incompatible buffer (hence min_size above)
    # Now, what happens if we do another thing with the same primary key?
    ConjectureData.for_buffer(buffer).draw(strategy)
```

Not sure how this will go, but it seems worth a try.

Reply:
This seemed to work. The only oddity is that the error message is less good if the fix is removed: it gives "FailedHealthCheck: It looks like your strategy is filtering out a lot of data" rather than the more explicit integrity error. I think this is just a side-effect of the fact that `_models_impl` uses `reject()`, which is caught by `@given`, so perhaps this is a feature rather than a bug. Anyway, this seems like the better solution.